arangod/Aql/WindowNode.cpp

```diff
@@ -0,0 +1,458 @@
+    TRI_ASSERT(rangeRegister < RegisterPlan::MaxRegisterId);
+    readableInputRegisters.insert(rangeRegister);
+  }
+
+  // calculate the aggregate registers
+  std::vector<std::pair<RegisterId, RegisterId>> aggregateRegisters;
+  calcAggregateRegisters(aggregateRegisters, readableInputRegisters, writeableOutputRegisters);
+
+  TRI_ASSERT(aggregateRegisters.size() == _aggregateVariables.size());
+
+  auto registerInfos = createRegisterInfos(std::move(readableInputRegisters),
+                                           std::move(writeableOutputRegisters));
+
+  std::vector<std::string> aggregateTypes;
+  std::transform(_aggregateVariables.begin(), _aggregateVariables.end(),
+                 std::back_inserter(aggregateTypes),
+                 [](auto& it) { return it.type; });
+  TRI_ASSERT(aggregateTypes.size() == _aggregateVariables.size());
+
+  auto executorInfos =
+      WindowExecutorInfos(_bounds, rangeRegister, std::move(aggregateTypes),
+                          std::move(aggregateRegisters), engine.getQuery().warnings(),
+                          &_plan->getAst()->query().vpackOptions());
+
+  if (_rangeVariable == nullptr && _bounds.unboundedPreceding()) {
+    return std::make_unique<ExecutionBlockImpl<AccuWindowExecutor>>(
+        &engine, this, std::move(registerInfos), std::move(executorInfos));
+  }
+  return std::make_unique<ExecutionBlockImpl<WindowExecutor>>(&engine, this,
+                                                              std::move(registerInfos),
+                                                              std::move(executorInfos));
+}
+
+/// @brief clone ExecutionNode recursively
+ExecutionNode* WindowNode::clone(ExecutionPlan* plan, bool withDependencies,
+                                 bool withProperties) const {
+  auto aggregateVariables = _aggregateVariables;
+
+  if (withProperties) {
+    // need to re-create all variables
+
+    aggregateVariables.clear();
+
+    for (auto& it : _aggregateVariables) {
+      auto out = plan->getAst()->variables()->createVariable(it.outVar);
+      auto in = plan->getAst()->variables()->createVariable(it.inVar);
+      aggregateVariables.emplace_back(AggregateVarInfo{out, in, it.type});
+    }
+  }
+
+  auto c = std::make_unique<WindowNode>(plan, _id, WindowBounds(_bounds),
+                                        _rangeVariable, aggregateVariables);
+
+  return cloneHelper(std::move(c), withDependencies, withProperties);
+}
+
+/// @brief getVariablesUsedHere, modifying the set in-place
+void WindowNode::getVariablesUsedHere(VarSet& vars) const {
+  if (_rangeVariable) {
+    vars.emplace(_rangeVariable);
+  }
+  for (auto const& p : _aggregateVariables) {
+    vars.emplace(p.inVar);
+  }
+}
+
+void WindowNode::setAggregateVariables(std::vector<AggregateVarInfo> const& aggregateVariables) {
+  _aggregateVariables = aggregateVariables;
+}
+
+/// @brief estimateCost
+CostEstimate WindowNode::estimateCost() const {
+  // FIXME fix
```

Comment:
For the range-based variant we will be out of luck, but can we try to add a cost estimation at least for the row-based variant?